src/metrics-collector/db/sqldb/sqldb.go

```diff
@@ -0,0 +1,153 @@
+package sqldb
+
+import (
+	"database/sql"
+	"encoding/json"
+	"strconv"
+
+	_ "github.com/lib/pq"
+	"github.com/pivotal-golang/lager"
+
+	"metrics-collector/config"
+	"metrics-collector/metrics"
+)
+
+const PostgresDriverName = "postgres"
+
+type SQLDB struct {
+	database *sql.DB
+	logger   lager.Logger
+	conf     *config.DbConfig
+}
+
+func NewSQLDB(conf *config.DbConfig, logger lager.Logger) (*SQLDB, error) {
+	sqldb := &SQLDB{}
+	sqldb.conf = conf
+	sqldb.logger = logger
+
+	var err error
+	sqldb.database, err = sql.Open(PostgresDriverName, conf.Url)
+	if err != nil {
+		logger.Error("Connect-to-database", err)
+		return nil, err
+	}
+
+	err = sqldb.initMetricsTable()
+	if err != nil {
+		logger.Error("initialize-metrics-table", err)
+		return nil, err
+	}
+
+	return sqldb, err
+}
+
+func (db *SQLDB) initMetricsTable() error {
+	_, err := db.database.Exec(`
+		CREATE TABLE IF NOT EXISTS applicationmetrics(
+			appid VARCHAR(255),
+			name  VARCHAR(100),
+			unit  VARCHAR(20),
+			timestamp BIGINT,
+			value VARCHAR
+		)
+	`)
+	return err
+}
+
+func (db *SQLDB) SaveMetric(metric *metrics.Metric) error {
+	value, err := json.Marshal(metric.Instances)
+	if err != nil {
+		db.logger.Error("marshal-instance-metrics", err, lager.Data{"metric": metric})
+		return err
+	}
+
+	query := "INSERT INTO applicationmetrics(appid, name, unit, timestamp, value) values($1, $2, $3, $4, $5)"
+	_, err = db.database.Exec(query, metric.AppId, metric.Name, metric.Unit, metric.TimeStamp, string(value))
+
+	if err != nil {
+		db.logger.Error("insert-metric", err, lager.Data{"query": query})
+	}
+
+	return err
+}
+
+func (db *SQLDB) RetrieveMetrics(appid string, name string, start int64, end int64) ([]*metrics.Metric, error) {
+	query := "SELECT unit, timestamp, value FROM applicationmetrics WHERE " +
+		" appid = '" + appid + "'" +
+		" AND name = '" + name + "'" +
+		" AND timestamp >= " + strconv.FormatInt(start, 10)
+
+	if end >= 0 {
+		query = query + " AND timestamp <= " + strconv.FormatInt(end, 10)
+	}
+
+	mtrcs := []*metrics.Metric{}
+
+	rows, err := db.database.Query(query)
+	if err != nil {
+		db.logger.Error("retrive-metrics-from-database", err, lager.Data{"query": query})
+		return mtrcs, err
+	}
+
+	defer rows.Close()
+
+	for rows.Next() {
+		var unit, value string
+		var timestamp int64
+
+		if err = rows.Scan(&unit, &timestamp, &value); err != nil {
+			db.logger.Error("scan-metric-from-search-result", err)
+		}
+
+		inst := []metrics.InstanceMetric{}
+		err = json.Unmarshal([]byte(value), &inst)
```

Comment:
It might be slightly more efficient to create a Decoder outside the loop and just call d.Decode rather than create one every time within the json.Unmarshal.